CONTRIBUTING.md

```diff
@@ -0,0 +1,607 @@
+
+
+### Other Code Formatting not handled by ClangFormat
+
+#### Control flow constructs
+
+Single line statements should be avoided in all cases. Use of brackets is
+mandatory for all control flow constructs (e.g. `if`, `for`, `while`, ...).
+
+#### Variable declarations
+
+C++ supports declaring / defining multiple variables using a single
+_type-specifier_. However, this is often very confusing as references and
+pointers exhibit strange behavior:
+
+```c++
+template <typename T> using pointer = T *;
+
+int *        x, y;  // x is a pointer, y is not
+pointer<int> x, y;  // both x and y are pointers
+```
+
+For this reason, __always__ declare each variable on a separate line, with its
+own _type-specifier_.
+
+### CMake coding style
+
+#### Whitespaces
+
+All alignment in CMake files should use four spaces.
+
+#### Use of macros vs functions
+
+Macros in CMake do not have a scope. This means that any variable set in this
+macro will be available to the whole project. In contrast, functions in CMake
+have local scope and therefore all set variables are local only. In general,
+wrap all piece of algorithms using temporary variables in a function and use
+macros to propagate variables to the whole project.
+
+#### Naming style
+
+All Ginkgo specific variables should be prefixed with a `GINKGO_` and all
+functions by `ginkgo_`.
+
+
+## Helper scripts
+
+To facilitate easy development within Ginkgo and to encourage coders and
+scientists who do not want get bogged down by the details of the Ginkgo library,
+but rather focus on writing the algorithms and the kernels, Ginkgo provides the
+developers with a few helper scripts.
+
+### Create a new algorithm
+
+A `create_new_algorithm.sh` script is available for developers to facilitate
+easy addition of new algorithms. The options it provides can be queried with
+
+```sh
+./create_new_algorithm.sh --help
+```
+The main objective of this script is to add files and boiler plate code for the
+new algorithm using a model and an instance of that model. For example, models
+can be any one of `factorization`, `matrix`, `preconditioner` or `solver`. For
+example to create a new solver named `my_solver` similar to `gmres`, you would
+set the `ModelType` to `solver` and set the `ModelName` to `gmres`. This would
+duplicate the core algorithm and kernels of the `gmres` algorithm and replace
+the naming to `my_solver`. Additionally, all the kernels of the new `my_solver`
+are marked as `GKO_NOT_IMPLEMENTED`. For easy navigation and `.txt` file is created
+in the folder where the script is run, which lists all the TODO's. These TODO's can
+also be found in the corresponding files.
+
+### Converting CUDA code to HIP code
+
+This script calls the `hipify` script provided by HIP converting the CUDA syntax
+to HIP syntax. Additionally, it also automatically replaces the instances of
+CUDA with HIP as appropriate. Hence, this script can be called on a Ginkgo CUDA
+file.
```

Comment:
should we also mention cuda2hip script here?

Reply:
This was actually explaining the cuda2hip script. I just forgot to mention the script name itself.